core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -226,48 +357,215 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
+    if (sourceCodec == NoCompressionCodec && targetCodec == NoCompressionCodec) {
+      // check the magic value
+      if (!hasMagicValue(magicValueToUse)) {
+        // Message format conversion
+        convertNonCompressedMessages(offsetCounter, compactedTopic, now, messageTimestampType, magicValueToUse)
+      } else {
+        // Do in-place validation, offset assignment and maybe set timestamp
+        validateNonCompressedMessagesAndAssignOffsetInPlace(offsetCounter, now, compactedTopic, messageTimestampType,
+          messageTimestampDiffMaxMs)
       }
-      buffer.reset()
-      this
+
     } else {
-      // We need to deep-iterate over the message-set if any of these are true:
-      // (i) messages are compressed
-      // (ii) the topic is configured with a target compression codec so we need to recompress regardless of original codec
-      val messages = this.internalIterator(isShallow = false).map(messageAndOffset => {
-        if (compactedTopic && !messageAndOffset.message.hasKey)
-          throw new InvalidMessageException("Compacted topic cannot accept message without key.")
-
-        messageAndOffset.message
+      // Deal with compressed messages
+      // We only need to do re-compression in one of the followings situation:
+      // 1. Source and target compression codec are different
+      // 2. When magic value to use is 0 because offsets need to be overwritten
+      // 3. When magic value to use is above 0, but some fields of inner messages need to be overwritten.
+      // 4. Message format conversion is needed.
+
+      // Re-compression situation 1 and 2
+      var requireReCompression = sourceCodec != targetCodec || magicValueToUse == Message.MagicValue_V0
+
+      var maxTimestamp = Message.NoTimestamp
+      val expectedRelativeOffset = new AtomicLong(0)
+      val validatedMessages = new ListBuffer[Message]
+      this.internalIterator(isShallow = false).foreach(messageAndOffset => {
+        val message = messageAndOffset.message
+        validateMessageKey(message, compactedTopic)
+        if (message.magic > Message.MagicValue_V0 && magicValueToUse > Message.MagicValue_V0) {
+          // Re-compression situation 3
+          // Validate the timestamp
+          validateTimestamp(message, now, messageTimestampType, messageTimestampDiffMaxMs)
+          // Check if we need to overwrite offset
+          if (messageAndOffset.offset != expectedRelativeOffset.getAndIncrement)
+            requireReCompression = true
+          maxTimestamp = math.max(maxTimestamp, message.timestamp)
+        }
+
+        // Re-compression situation 4
+        if (message.magic != magicValueToUse)
+          requireReCompression = true
+
+        validatedMessages += message.toFormatVersion(magicValueToUse)
       })
-      new ByteBufferMessageSet(compressionCodec = targetCodec, offsetCounter = offsetCounter, messages = messages.toBuffer:_*)
+
+      if (requireReCompression) {
```

Comment:
requireReCompression is a bit confusing. For example, requireReCompression will be true if the source data is compressed and the broker requires no compression. Perhaps we can rename it to sth like inPlaceConversion and negate the test?